src/cryptonote_basic/cryptonote_format_utils.cpp

```diff
@@ -866,45 +889,114 @@ namespace cryptonote
     return res;
   }
   //---------------------------------------------------------------
-  bool is_out_to_acc(const account_keys& acc, const txout_to_key& out_key, const crypto::public_key& tx_pub_key, const std::vector<crypto::public_key>& additional_tx_pub_keys, size_t output_index)
+  void set_tx_out(const uint64_t amount, const crypto::public_key output_public_key, const bool use_view_tags, const crypto::view_tag view_tag, tx_out &out)
+  {
+    out.amount = amount;
+    if (use_view_tags)
+    {
+      txout_to_tagged_key ttk;
+      ttk.key = output_public_key;
+      ttk.view_tag = view_tag;
+      out.target = ttk;
+    }
+    else
+    {
+      txout_to_key tk;
+      tk.key = output_public_key;
+      out.target = tk;
+    }
+  }
+  //---------------------------------------------------------------
+  bool check_output_types(const transaction& tx, const uint8_t hf_version)
+  {
+    for (const auto &o: tx.vout)
+    {
+      if (hf_version >= HF_VERSION_VIEW_TAGS)
+      {
+        // from v15, require outputs have view tags
+        CHECK_AND_ASSERT_MES(o.target.type() == typeid(txout_to_tagged_key), false, "wrong variant type: "
+          << o.target.type().name() << ", expected txout_to_tagged_key in transaction id=" << get_transaction_hash(tx));
+      }
+      else
+      {
+        // require outputs to be of type txout_to_key
+        CHECK_AND_ASSERT_MES(o.target.type() == typeid(txout_to_key), false, "wrong variant type: "
+          << o.target.type().name() << ", expected txout_to_key in transaction id=" << get_transaction_hash(tx));
+      }
+    }
+    return true;
+  }
```

Comment:
After our "two hardforks in rapid succession" discussion with a "grace period" as result this is one point where it will get a little more complicated, right?

Reply:
Yep, precisely. I plan to change to:

```
if (hf_version < HF_VERSION_VIEW_TAGS)
   require tx_out_to_key
else if (hf_version > HF_VERSION_VIEW_TAGS)
   require tx_out_to_tagged_key
else if (hf_version == HF_VERSION_VIEW_TAGS)
   require tx_out_to_key || tx_out_to_tagged_key
```

I'll also update `hardforks.cpp` to have both a version 15 and version 16, and will add some tests for this behavior.

Probably makes sense for the `== HF_VERSION_VIEW_TAGS` case to require all outputs in the tx to be of the same type too, just so it doesn't allow for weird deviations from expected behavior

Reply:
Good idea to check for the same type for all outputs of a transaction.